control-plane/csi-controller/src/controller.rs

```diff
@@ -0,0 +1,774 @@
+                    // Make sure volume is published at the same node.
+                    if node_id != node {
+                        let m = format!(
+                            "Volume {} already published on a different node: {}",
+                            volume_id, node,
+                        );
+                        error!("{}", m);
+                        return Err(Status::failed_precondition(m));
+                    }
+
+                    debug!("Volume {} already published at {}", volume_id, uri);
+                    uri
+                } else {
+                    let m = format!(
+                        "Volume {} reports no info about its publishing status",
+                        volume_id
+                    );
+                    error!("{}", m);
+                    return Err(Status::internal(m));
+                }
+            } else {
+                // Volume is not published.
+                let v = MayastorApiClient::get_client()
+                    .publish_volume(&volume_id, &node_id, protocol)
+                    .await?;
+
+                if let Some((node, uri)) = get_volume_share_location(&v) {
+                    debug!(
+                        "Volume {} successfully published on node {} via {}",
+                        volume_id, node, uri
+                    );
+                    uri
+                } else {
+                    let m = format!(
+                        "Volume {} has been successfully published but URI is available",
+                        volume_id
+                    );
+                    error!("{}", m);
+                    return Err(Status::internal(m));
+                }
+            }
+        } else {
+            let m = format!("Volume {} is missing current state", volume_id);
+            error!("{}", m);
+            return Err(Status::internal(m));
+        };
+
+        // Prepare the context for the Mayastor Node CSI plugin.
+        let mut publish_context = HashMap::new();
+        publish_context.insert("uri".to_string(), uri);
+
+        if let Some(io_timeout) = args.volume_context.get(volume_opts::IO_TIMEOUT) {
+            publish_context.insert(volume_opts::IO_TIMEOUT.to_string(), io_timeout.to_string());
+        }
+
+        debug!(
+            "Publish context for volume {}: {:?}",
+            volume_id, publish_context
+        );
+        Ok(Response::new(ControllerPublishVolumeResponse {
+            publish_context,
+        }))
+    }
+
+    #[instrument]
+    async fn controller_unpublish_volume(
+        &self,
+        request: tonic::Request<ControllerUnpublishVolumeRequest>,
+    ) -> Result<tonic::Response<ControllerUnpublishVolumeResponse>, tonic::Status> {
+        let args = request.into_inner();
+
+        debug!("Request to unpublish volume: {:?}", args);
+        // Check if target volume exists.
+        let volume = match MayastorApiClient::get_client()
+            .get_volume(&args.volume_id)
```

Comment:
If I understood you reasoning with the 404, I think this is one of the cases where we'd want to use volumes/{id}, rather than volumes?id={id}, because the volume should indeed exist, and the 404 here is an error?